snapshot/iterator_test.go

```diff
@@ -0,0 +1,193 @@
+// Modifications Copyright 2021 The klaytn Authors
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+//
+// This file is derived from core/state/snapshot/iterator_test.go (2021/10/21).
+// Modified and improved for the klaytn development.
+
+package snapshot
+
+import (
+	"bytes"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"testing"
+
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/storage/database"
+)
+
+// TODO-snapshot remove the following function after porting diffToDisk method
+func createTestDiskLayer(db database.DBManager, accounts map[common.Hash][]byte, storages map[common.Hash]map[common.Hash][]byte) *diskLayer {
+	batch := db.NewSnapshotDBBatch()
+
+	for hash, data := range accounts {
+		batch.WriteAccountSnapshot(hash, data)
+		batch.Write()
+		batch.Reset()
+	}
+
+	for accountHash, storage := range storages {
+		for storageHash, data := range storage {
+			if len(data) > 0 {
+				batch.WriteStorageSnapshot(accountHash, storageHash, data)
+			} else {
+				batch.DeleteStorageSnapshot(accountHash, storageHash)
+			}
+			batch.Write()
+			batch.Reset()
+		}
+	}
+
+	return &diskLayer{
+		diskdb: db,
+	}
+}
+
+// TestAccountIteratorBasics tests some simple single-layer(diff and disk) iteration
+func TestAccountIteratorBasics(t *testing.T) {
+	var (
+		destructs = make(map[common.Hash]struct{})
+		accounts  = make(map[common.Hash][]byte)
+		storage   = make(map[common.Hash]map[common.Hash][]byte)
+	)
+	// Fill up a parent
+	for i := 0; i < 100; i++ {
+		h := randomHash()
+		data := randomAccount()
+
+		accounts[h] = data
+		if rand.Intn(4) == 0 {
+			destructs[h] = struct{}{}
+		}
+		if rand.Intn(2) == 0 {
```

Comment:
What happens both if statement is true?
destructs[h] and storage[h] will be set. Is this okay?

Reply:
In this PR, the `destructs` are not used actually. But, both fields can be set at the same time. A contract can be updated and destructed at the same block. But, for now only accounts and storage fields are used for this test case.

Reply:
got it. Thanks.